File: python/Q17_letter_combinations_of_a_phone_number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#


# @lc code=start
class Solution:
    def backtrack(self, index: int, stack: list[str]) -> None:
        if index == self.length:
            self.result.append("".join(stack))
            return

        for c in self.mapping[self.digits[index]]:
            stack.append(c)
            self.backtrack(index + 1, stack)
            stack.pop()

    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        self.digits: str = digits
        self.length: int = len(digits)
        self.mapping: dict[str, list[str]] = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        self.result: list[str] = []

        self.backtrack(0, [])

        return self.result


# @lc code=end

File: rust/src/leetcode/Q17_letter_combinations_of_a_phone_number.rs
/*
 * @lc app=leetcode id=17 lang=rust
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
impl Solution {
    fn backtrack(
        digits: &Vec<usize>,
        mapping: &[&[char]],
        index: usize,
        current: &mut Vec<char>,
        result: &mut Vec<String>,
    ) {
        if index == digits.len() {
            result.push(current.iter().collect());
            return;
        }

        for &c in mapping[digits[index]] {
            current.push(c);

            Self::backtrack(digits, mapping, index + 1, current, result);

            current.pop();
        }
    }

    pub fn letter_combinations(digits: String) -> Vec<String> {
        if digits.is_empty() {
            return Vec::new();
        }

        let mapping: [&[char]; 10] = [
            &[],                   // 0
            &[],                   // 1
            &['a', 'b', 'c'],      // 2
            &['d', 'e', 'f'],      // 3
            &['g', 'h', 'i'],      // 4
            &['j', 'k', 'l'],      // 5
            &['m', 'n', 'o'],      // 6
            &['p', 'q', 'r', 's'], // 7
            &['t', 'u', 'v'],      // 8
            &['w', 'x', 'y', 'z'], // 9
        ];

        let digits: Vec<usize> = digits
            .chars()
            .map(|d| d.to_digit(10).unwrap() as usize)
            .collect();
        let mut current: Vec<char> = Vec::new();
        let mut result: Vec<String> = Vec::new();

        Self::backtrack(&digits, &mapping, 0, &mut current, &mut result);

        result
    }
}
// @lc code=end
